onnxruntime/core/providers/cuda/cuda_execution_provider.cc

```diff
@@ -2203,23 +2203,19 @@ static Status RegisterCudaKernels(KernelRegistry& kernel_registry) {
 
 }  // namespace cuda
 
-static std::shared_ptr<KernelRegistry>& CudaKernelRegistry() {
-  // static local variable ensures thread-safe initialization
-  static std::shared_ptr<KernelRegistry> cuda_kernel_registry = []() {
-    std::shared_ptr<KernelRegistry> registry = KernelRegistry::Create();
-    ORT_THROW_IF_ERROR(cuda::RegisterCudaKernels(*registry));
-    return registry;
-  }();
-
-  return cuda_kernel_registry;
+static std::shared_ptr<KernelRegistry> s_kernel_registry;
+
+void InitializeRegistry() {
+  s_kernel_registry = KernelRegistry::Create();
```

Comment:
what about thread-safety?

Reply:
It's only ever called by the shared provider loader code, the same place the Shutdown() code is. We might want to make ProviderLibrary::Get() have a critical section in it as it could be called by multiple threads calling SessionOptionsAppendExecutionProvider_* (not sure if that API is thread safe, many at that level are not)

Reply:
that sounds like a good idea. I suppose Provider::Initialize() and Provider::Shutdown() should be in a critical section/protected by the same mutex.

Reply:
It should be more thread safe than ever now for loading the providers.

Reply:
I had to remove the Unload() mutex lock on Unload sadly, see the comments in the code.